### ISSUE TITLE
Wireless Scene Switch 3-gang Zigbee TS0043 _TZ3000_w8jwkczz

### ISSUE THREAD

**Issue by kvnnap** [dresden-elektronik/deconz-rest-plugin#5552]
## Device

- Product name: Scene Switch (Powered by Tuya)
- Manufacturer: _TZ3000_w8jwkczz
- Model identifier: TS0043
- Device type :
  - Switch
  Front:
  ![front](https://user-images.githubusercontent.com/16206628/144703776-c6e87e6b-895a-4487-a56c-6df0235227dd.jpg)
  Back:
  ![back](https://user-images.githubusercontent.com/16206628/144703789-1e500b8b-5f81-4d02-a091-ea97e77151be.jpg)

## Screenshots

![switch-general](https://user-images.githubusercontent.com/16206628/144704150-85bf5961-d0c2-4f13-99ee-7279719d36f0.png)

![switch-node-info](https://user-images.githubusercontent.com/16206628/144704153-8e7bcd92-0212-4c80-98a1-36ea54e2d7aa.png)

### Basic
![switch-01-Basic](https://user-images.githubusercontent.com/16206628/144704170-1b13c023-c86f-4cc4-a8f4-108ed65426f7.png)

### Power Configuration
![switch-01-Power](https://user-images.githubusercontent.com/16206628/144704177-aab11a18-bb04-4108-8fe3-5a42cca9cd28.png)

### On/Off
![switch-01-On-Off](https://user-images.githubusercontent.com/16206628/144704188-b1a2d736-136f-4f3a-a06b-29a91cd9e99f.png)

### OTAU (Grey)
![switch-01-OTAU](https://user-images.githubusercontent.com/16206628/144704191-6c74e10a-6816-41b5-9e01-80b6307ad411.png)

### Time (Grey)
Shows nothing.

### On/Off (Grey)
![switch-01-On-Off-gray](https://user-images.githubusercontent.com/16206628/144704212-d2178441-0bd0-4f28-9be7-00416ec0bd44.png)

### Power Configuration
![switch-02-Power](https://user-images.githubusercontent.com/16206628/144704218-8ad25032-26d4-49da-9b5a-07f296ecfa84.png)

### On/Off
![switch-02-On-Off](https://user-images.githubusercontent.com/16206628/144704224-3c73a3a0-01f2-48da-8820-4e38f59d147f.png)

### On/Off (Grey)
![switch-02-On-Off-gray](https://user-images.githubusercontent.com/16206628/144704230-ed1b14ad-687c-4751-9bd7-cdbcea2b4842.png)

### Power Configuration
![switch-03-Power](https://user-images.githubusercontent.com/16206628/144704233-cbba8f5f-4afd-47c8-a511-27281bc981ec.png)

### On/Off
![switch-03-On-Off](https://user-images.githubusercontent.com/16206628/144704237-f1ceece9-49cf-4f10-908e-8cf236a42b6f.png)

### On/Off (Grey)
![switch-03-On-Off-gray](https://user-images.githubusercontent.com/16206628/144704241-1916046d-05a0-40ec-9ada-0702f64bb117.png)

## Attempt

I have attempted to imitate other code changes, such as this PR [here](https://github.com/dresden-elektronik/deconz-rest-plugin/pull/4756/files) and compiled, tested and prepared a commit [here](https://github.com/kvnnap/deconz-rest-plugin/commit/9e4d97cbf4cab1f5e0ca8b9244179f7254d3f657). The device buttons work. I see the correct codes when a call to the REST API is made. However, I still see 3 On/Off entries under Lights in Phoscon. The resulting json and screenshots are below:

Correct switch addition:

![phoscon-switches](https://user-images.githubusercontent.com/16206628/144705566-c8860180-704a-4c94-b4f7-87923e59a017.png)

```json
    "3": {
        "config": {
            "battery": null,
            "on": true,
            "reachable": true
        },
        "ep": 1,
        "etag": "7b783e20cf59c61f25ff3e2841af53a2",
        "lastannounced": "2021-12-04T10:01:50Z",
        "lastseen": "2021-12-04T10:01Z",
        "manufacturername": "_TZ3000_w8jwkczz",
        "mode": 1,
        "modelid": "TS0043",
        "name": "Switch",
        "state": {
            "buttonevent": 3002,
            "lastupdated": "2021-12-04T09:56:12.904"
        },
        "type": "ZHASwitch",
        "uniqueid": "80:4b:50:ff:fe:06:9a:e6-01-0006"
    }
```

Incorrect light additions:

![phoscon-lights](https://user-images.githubusercontent.com/16206628/144705574-775ed0e6-f6fe-42ea-aa45-67f7f4a124f8.png)

```json
    "8": {
        "etag": "0f4db69999cf0585054066ba8287affb",
        "hascolor": false,
        "lastannounced": "2021-12-04T10:01:50Z",
        "lastseen": "2021-12-04T10:01Z",
        "manufacturername": "_TZ3000_w8jwkczz",
        "modelid": "TS0043",
        "name": "On/Off light 8",
        "state": {
            "alert": "none",
            "on": false,
            "reachable": true
        },
        "swversion": null,
        "type": "On/Off light",
        "uniqueid": "80:4b:50:ff:fe:06:9a:e6-01"
    },
    "9": {
        "etag": "e3443691a63e2147e780d80e469dc850",
        "hascolor": false,
        "lastannounced": "2021-12-04T10:01:50Z",
        "lastseen": "2021-12-04T10:01Z",
        "manufacturername": "_TZ3000_w8jwkczz",
        "modelid": "TS0043",
        "name": "On/Off light 9",
        "state": {
            "alert": "none",
            "on": false,
            "reachable": true
        },
        "swversion": null,
        "type": "On/Off light",
        "uniqueid": "80:4b:50:ff:fe:06:9a:e6-02"
    },
    "10": {
        "etag": "0555947fec87e0ea2599a6ad5d17e405",
        "hascolor": false,
        "lastannounced": "2021-12-04T10:01:50Z",
        "lastseen": "2021-12-04T10:01Z",
        "manufacturername": "_TZ3000_w8jwkczz",
        "modelid": "TS0043",
        "name": "On/Off light 10",
        "state": {
            "alert": "none",
            "on": false,
            "reachable": true
        },
        "swversion": null,
        "type": "On/Off light",
        "uniqueid": "80:4b:50:ff:fe:06:9a:e6-03"
    }
```

## Environment and Other

The software runs on a Rasperry Pi 4b inside a docker container. To compile the change, the following Dockerfile was produced:

```dockerfile
FROM debian:10.11-slim

ARG MYUSER=deconz

RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
    && apt-get -y dist-upgrade \
    && apt-get -y install wget gnupg2 g++ make qtdeclarative5-dev \
    && wget -O - http://phoscon.de/apt/deconz.pub.key | apt-key add - \
    && echo "deb http://phoscon.de/apt/deconz $(cat /etc/os-release | grep -oP \
      '(?<=VERSION_CODENAME\=).*')-beta main" > /etc/apt/sources.list.d/deconz.list \
    && apt-get update \
    && apt-get -y dist-upgrade \
    && apt-get -y install deconz-dev \
    && useradd -ms /bin/bash $MYUSER \
    && apt-get -y install --no-install-recommends sudo \
    && echo $MYUSER ALL=\(root\) NOPASSWD:ALL > /etc/sudoers.d/$MYUSER \
    && chmod 0440 /etc/sudoers.d/$MYUSER

USER $MYUSER
WORKDIR /home/$MYUSER
```

and this Dockerfile was used to test the change after libde_rest_plugin.so was compiled:

```dockerfile
FROM deconzcommunity/deconz:latest
COPY libde_rest_plugin.so /usr/share/deCONZ/plugins
COPY button_maps.json /usr/share/deCONZ/devices
```

**Comment by Mimiix** [comment]
@Smanar

**Comment by Smanar** [comment]
It's normal.
Just ignore them for the moment, need to cxode something to prevent their creation.

**Comment by kvnnap** [comment]
ok, will test the code as soon as there is an update.

**Comment by Smanar** [comment]
I think the next update will be a DDF file and need more time for battery switch (the code is not ready yet)

**Comment by kvnnap** [comment]
Keeping issue alive. @Smanar Any updates on this?

**Comment by Smanar** [comment]
Nope, for the moment all PR are blocked.
The official method is using the DDF file, but the deconz core don't support battery switches yet.
So for the moment all is waiting list.

**Comment by kvnnap** [comment]
Keeping issue alive.

**Comment by kvnnap** [comment]
Hi @Smanar 

Is there any timeline about this and is there anywhere I can follow updates about support for battery switches?

I am asking so I can get an idea about when this issue will be tackled.

**Comment by Mimiix** [comment]
@kvnnap please check #5733

**Comment by kvnnap** [comment]
Keeping this alive..

**Comment by kvnnap** [comment]
keeping this alive

**Comment by kvnnap** [comment]
keeping this alive

**Comment by Smanar** [comment]
Hello.
It's old, but If I right the code is working, you justt need to want to remove the useless entries ?

I think it's possible using DDF.
We will use them during the creation and let the legacy code make his work after.

https://github.com/dresden-elektronik/deconz-rest-plugin/wiki/DDF-cheat-sheet

Just select "edit DDF", and remove useless light entry, or just add a ZHASwitch one. If it don't work can you copy paste your DDF ?

**Comment by kvnnap** [comment]
Hi @Smanar,

Been busy, will try later today and let you know.

Thanks!

**Comment by kvnnap** [comment]
Hi @Smanar,

I tried editing the DDF and removing the useless light entries, however they still show in Phoscon Web. Then I tried removing the lights from Phoscon Web but it removed the switch too.

Not sure how to proceed.

**Comment by Smanar** [comment]
Are you using the GUI or text editor ? The easier way is reincluding the device with a working DDF, don't forget to set the status to "Gold"
Can you show your final DDF ?

And yes it s something normal, you can't remove a "part of a device", it s all or nothing.

**Comment by kvnnap** [comment]
I was using the GUI and I set it to Silver (but allowed silver DDFs).  I just tested with status Gold and I have the same result. I also tried removing the device and adding it back after saving this DDF. Here's the final DDF:

```json
{
  "schema": "devcap1.schema.json",
  "manufacturername": "_TZ3000_w8jwkczz",
  "modelid": "TS0043",
  "product": "TS0043",
  "sleeper": false,
  "status": "Gold",
  "path": "/devices/ts0043.json",
  "subdevices": [
    {
      "type": "$TYPE_SWITCH",
      "restapi": "/sensors",
      "uuid": [
        "$address.ext",
        "0x01",
        "0x0006"
      ],
      "fingerprint": {
        "profile": "0x0104",
        "device": "0x0000",
        "endpoint": "0x01",
        "in": [
          "0x0000",
          "0x0001"
        ],
        "out": [
          "0x0006"
        ]
      },
      "items": [
        {
          "name": "attr/id"
        },
        {
          "name": "attr/lastannounced"
        },
        {
          "name": "attr/lastseen"
        },
        {
          "name": "attr/manufacturername"
        },
        {
          "name": "attr/modelid"
        },
        {
          "name": "attr/name"
        },
        {
          "name": "attr/swversion"
        },
        {
          "name": "attr/type"
        },
        {
          "name": "attr/uniqueid"
        },
        {
          "name": "config/battery",
          "description": "The current device battery level in 0&ndash;100&thinsp;%.",
          "default": 0
        },
        {
          "name": "config/on"
        },
        {
          "name": "config/reachable"
        },
        {
          "name": "state/buttonevent",
          "description": "The last received button event."
        },
        {
          "name": "state/lastupdated"
        }
      ]
    }
  ],
  "bindings": [
    {
      "bind": "unicast",
      "src.ep": 1,
      "dst.ep": 1,
      "cl": "0x0001",
      "report": [
        {
          "at": "0x0021",
          "dt": "0x20",
          "min": 300,
          "max": 2700,
          "change": "0x00000001"
        }
      ]
    },
    {
      "bind": "unicast",
      "src.ep": 2,
      "dst.ep": 1,
      "cl": "0x0001",
      "report": [
        {
          "at": "0x0021",
          "dt": "0x20",
          "min": 300,
          "max": 2700,
          "change": "0x00000001"
        }
      ]
    },
    {
      "bind": "unicast",
      "src.ep": 3,
      "dst.ep": 1,
      "cl": "0x0001",
      "report": [
        {
          "at": "0x0021",
          "dt": "0x20",
          "min": 300,
          "max": 2700,
          "change": "0x00000001"
        }
      ]
    }
  ]
}
```

**Comment by Smanar** [comment]
So all seem fine for me ...
And if you try to include the device you have 1 sensor and still 3 lights entries.
Will ask to others devs.

**Comment by kvnnap** [comment]
Correct, the web interface still shows 1 switch and 3 light entries. 

Current deCONZ version: 2.15.03

I am using this docker image:  deconzcommunity/deconz:latest

![image](https://user-images.githubusercontent.com/16206628/167263444-feb35277-ad63-4ba5-a5a8-8566ca2489c4.png)

**Comment by Smanar** [comment]
I m asking to other devs, Not sure at 100% the legacy code is bypassed by DDF, but the result is still strange.
You have a way to be sure the DDF is used ? because it s hard to use DDF on docker OS, you can see it on the GUI (on the editor window title when making "edit DDF") or on logs ?

**Comment by kvnnap** [comment]
I restarted the container and noticed that the logs show that it is loaded:

![image](https://user-images.githubusercontent.com/16206628/167266670-782d6af8-d0f2-43c3-a789-46f2e7fbd2f5.png)

Also, when I click Edit DDF on the switch I get this:

![image](https://user-images.githubusercontent.com/16206628/167266710-60f53914-c1cf-47e7-9724-99e47c8f1f95.png)

Phoscon Web still shows:

![image](https://user-images.githubusercontent.com/16206628/167266731-a98c27da-133c-4fdd-8c62-1b846b16ee3d.png)

and

![image](https://user-images.githubusercontent.com/16206628/167266760-5a572037-0064-4ab9-8fb8-14b8486fcc88.png)


I also tried reverting the docker container to a version where I have no switch setup and restarting this whole setup, but it still gives this result.

**Comment by Mimiix** [comment]
Phoscon will never show the correct result as they have to add support on their own.

Please use the rest api and websocket to determine if the  behavior is correct.

**Comment by Smanar** [comment]
He have give the API extract on the first post, with all json.

For me it's realy an issue on DDF core

**Comment by kvnnap** [comment]
I checked the API to make sure and the extra lights are still listed:

![image](https://user-images.githubusercontent.com/16206628/167286945-3d9c31bf-1eb5-40e5-aae9-a5b2b86d172a.png)

**Comment by kvnnap** [comment]
I have also noticed that I am unable to get the button events (buttonevent is null). Prior to DDF I had changed the button_maps.json file (see [here](https://github.com/kvnnap/deconz-rest-plugin/commit/9e4d97cbf4cab1f5e0ca8b9244179f7254d3f657)) in order to make it work. How does it work now with DDF?

```json
{
    "config": {
        "battery": 0,
        "on": true,
        "reachable": true
    },
    "ep": 2,
    "etag": "72e466e434e9860936a90f40f4328e37",
    "lastannounced": "2022-05-08T08:52:49Z",
    "lastseen": "2022-05-08T08:52Z",
    "manufacturername": "_TZ3000_w8jwkczz",
    "mode": 1,
    "modelid": "TS0043",
    "name": "Switch 7",
    "state": {
        "buttonevent": null,
        "lastupdated": "none"
    },
    "type": "ZHASwitch",
    "uniqueid": "80:4b:50:ff:fe:06:9a:e6-02-0006"
}
```

**Comment by Smanar** [comment]
I m seing this PR https://github.com/dresden-elektronik/deconz-rest-plugin/pull/6021

I readly don't see what do the code but seem usefull according to description


> Since DDF devices and their light and sensor resources are created async, the rule triggers can't be indexed in the plugin constructor.
> 
> The PR delays/debounces the indexing so that all devices are handled properly. This is also a bit faster since prior for each device useless container calls where made during startup.

**Comment by kvnnap** [comment]
Is this PR going to be included in the upcoming release 2.15.04? If so I will wait and test when this is out.

**Comment by Smanar** [comment]
Yep, it's already merged.
